relay-server/src/actors/upstream.rs

```diff
@@ -284,43 +319,114 @@ impl UpstreamRelay {
+            network_status_backoff: RetryBackoff::new(config.http_max_retry_interval()),
             max_inflight_requests: config.max_concurrent_requests(),
             auth_state: AuthState::Unknown,
             num_inflight_requests: 0,
             high_prio_requests: VecDeque::new(),
             low_prio_requests: VecDeque::new(),
             first_error: None,
+            authentication_denied: false,
+            is_waiting_to_authenticate: false,
             config,
         }
     }
 
+    /// Predicate, checks if a Relay performs authentication
+    fn authenticates(&self) -> bool {
+        // only managed mode relays perform authentication
+        self.config.relay_mode() == RelayMode::Managed
+    }
+
+    /// Predicate, checks if a Relay does re-authentication
+    fn re_authenticates(&self) -> bool {
+        // only relays that authenticate also re-authenticate
+        self.authenticates()
+            // only relays the have a configured auth-interval reauthenticate
+            && self.config.http_auth_interval().is_some()
+            // processing relays do NOT re-authenticate
+            && !self.config.processing_enabled()
+            // the upstream did not ban us explicitly from trying to re-authenticate
+            && !self.authentication_denied
+    }
+
+    /// Predicate, checks if we are in an network outage situation
+    fn is_network_outage(&self) -> bool {
+        if let Some(first_error) = self.first_error {
+            let now = Instant::now();
+            first_error + self.config.network_outage_grace_period() < now
+        } else {
+            false
+        }
+    }
+
+    /// Returns an error message if an authentication is prohibited in this state and
+    /// None if it can authenticate.
+    fn get_auth_state_error(&self) -> Option<&'static str> {
+        if !self.authenticates() {
+            Some("Upstream actor trying to authenticate although it is not supposed to.")
+        } else if self.auth_state == AuthState::Registered && !self.re_authenticates() {
+            Some("Upstream actor trying to re-authenticate although it is not supposed to.")
+        } else if self.authentication_denied {
+            Some("Upstream actor trying to authenticate after authentication was denied.")
+        } else {
+            // Ok to authenticate
+            None
+        }
+    }
+
+    /// Predicate, checks if Upstream is in the process of authentication.
+    /// That is it checks if an authentication is running at the moment or an
+    /// authentication has just failed and Upstream is waiting for the backoff
+    /// interval to finish in order to try again.
+    //fn is_authenticating_now(&self) -> bool {
+    //    self.backoff.started()
+    //}
+
+    /// called when a network error is detected in order to keep track
+    fn remember_network_error(&mut self) {
+        if self.first_error == None {
+            self.first_error = Some(Instant::now())
+        }
```

Comment:
_nit_: This can be written with `get_or_insert_with`:

```suggestion
        self.first_error.get_or_insert_with(Instant::now);
```

I would consider to return first error from this function, so that it can be used immediately for checking the outage.